common/djangoapps/third_party_auth/provider.py

```diff
@@ -29,21 +30,43 @@ def _enabled_providers(cls):
         """
         Helper method that returns a generator used to iterate over all providers
         of the current site.
+
+        Provider configurations are partitioned on site + some key (backend
+        name in the case of OAuth, slug for SAML, and consumer key for LTI).
         """
-        oauth2_backend_names = OAuth2ProviderConfig.key_values('backend_name', flat=True)
-        for oauth2_backend_name in oauth2_backend_names:
-            provider = OAuth2ProviderConfig.current(oauth2_backend_name)
-            if provider.enabled_for_current_site and provider.backend_name in _PSA_OAUTH2_BACKENDS:
+        def current_partitioned(prov_cls, filters=None):
+            """
+            Like ConfigurationModel.current() but partitions on KEY_FIELDS and
+            allows additional filters dict. (Does not include caching.)
+            """
+            def partition_key(obj):
+                """Compute the partition key for this object."""
+                return tuple(getattr(obj, k) for k in prov_cls.KEY_FIELDS)
+
+            configs = prov_cls.objects.filter(**filters) if filters else prov_cls.objects.all()
+            # Get the most recent record in each partition
+            return unique_everseen(configs.order_by('-change_date'), partition_key)
+
+        site = Site.objects.get_current(get_current_request())
+
+        def enabled_in_site(prov_cls, filters=None):
+            """
+            Generator for all current configs of this class that are enabled
+            and for this site.
+            """
+            for config in current_partitioned(prov_cls, dict(site=site, **(filters or {}))):
+                if config.enabled:
+                    yield config
+
+        for provider in enabled_in_site(OAuth2ProviderConfig, {'backend_name__in': _PSA_OAUTH2_BACKENDS}):
```

Comment:
Can you please add more comments for this loop? Like `First, we should get all the ID providers using oauth2`